tortoise/verifying_tortoise.go

```diff
@@ -305,89 +265,158 @@ func (t *turtle) checkBlockAndGetLocalOpinion(
-	voteMap = make(map[types.BlockID]vec, len(candidateBlocks))
-	if opinionVec == nil {
-		// nil means abstain, i.e., we have no opinion on blocks in this layer
-		for _, b := range candidateBlocks {
-			voteMap[b] = abstain
-		}
-		return
+func (t *turtle) computeLocalOpinion(ctx *tcontext, lid types.LayerID) (map[types.BlockID]vec, error) {
+	bids, err := t.getLayerBlocksIDs(ctx, lid)
+	if err != nil {
+		return nil, err
 	}
-
-	// add support for all blocks in input vector
-	for _, b := range opinionVec {
-		voteMap[b] = support
+	supported, err := t.layerOpinionVector(ctx, lid)
+	if err != nil {
+		return nil, err
 	}
+	return t.layerVoteVector(bids, supported), nil
+}
 
-	// vote against all layer blocks not in input vector
-	for _, b := range candidateBlocks {
-		if _, ok := voteMap[b]; !ok {
-			voteMap[b] = against
-		}
+func (t *turtle) layerVoteVector(bids, supported []types.BlockID) map[types.BlockID]vec {
+	opinion := map[types.BlockID]vec{}
+	defaultOpinion := against
+	if supported == nil { //nolint
+		defaultOpinion = abstain
 	}
-	return
+	for _, on := range bids {
+		opinion[on] = defaultOpinion
+	}
+	for _, on := range supported { //nolint
+		opinion[on] = support
+	}
+	return opinion
 }
 
-// BaseBlock finds and returns a good block from a recent layer that's suitable to serve as a base block for opinions
-// for newly-constructed blocks. Also includes vectors of exceptions, where the current local opinion differs from
-// (or is newer than) that of the base block.
+// BaseBlock selects a base block from sliding window based on a following priorities in order:
+// - choose good block
+// - choose block with the least difference to the local opinion
+// - choose block from higher layer
+// - otherwise deterministically select block with lowest id.
 func (t *turtle) BaseBlock(ctx context.Context) (types.BlockID, [][]types.BlockID, error) {
-	logger := t.logger.WithContext(ctx)
-
-	// look at good blocks backwards from most recent processed layer to find a suitable base block
-	// TODO: optimize by, e.g., trying to minimize the size of the exception list (rather than first match)
-	// see https://github.com/spacemeshos/go-spacemesh/issues/2402
-	for layerID := t.Last; layerID.After(t.LastEvicted); layerID = layerID.Sub(1) {
-		logger := logger.WithFields(
-			log.FieldNamed("last_layer", t.Last),
-			log.FieldNamed("last_evicted", t.LastEvicted),
-			log.FieldNamed("base_block_candidate_layer", layerID))
-		for blockID, opinion := range t.BlockOpinionsByLayer[layerID] {
-			if _, ok := t.GoodBlocksIndex[blockID]; !ok {
-				logger.With().Info("not considering block not marked good as base block candidate", blockID)
-				continue
-			}
+	var (
+		tctx          = wrapContext(ctx)
+		logger        = t.logger.WithContext(ctx)
+		disagreements = map[types.BlockID]types.LayerID{}
+		choices       []types.BlockID // choices from the best to the least bad
```

Comment:
good idea, will consider to implement

the problem with choosing base block out of sliding window though is that everyone else will evict it,so that block that will receive base block outside of sliding window is useless

Reply:
i think i will keep it as is for simplicity. as i mentioned we don't want blocks from that are evicted, bad block that is not evicted will allow full tortoise to make progress, evicted blocks will stall the network eventually. so this is only about the performance, and it is sufficient for now